scipy/stats/stats.py

```diff
@@ -4275,6 +4276,164 @@ def mannwhitneyu(x, y, use_continuity=True):
+    exact : {True, False, 'auto'}, optional
+        Whether an exact test should be performed on the data. See notes for
+        default behavior.
+    alternative : str, optional
+        Whether a two-tailed or one-tailed test should be performed on the
+        supplied vectors. Arguments are 'two-tailed', 'less', and 'greater'.
+        Default is 'two-tailed'.
+
+    Returns
+    -------
+    Bunch object with following attributes:
+        statistic : float
+            The test statistic.
+        pvalue : float
+            The pvalue of the test.
+        exact : bool
+            Indicates if an exact pvalue was calculated.
+        alternative : str
+            Describes the alternative hypothesis.
+        u1 : float
+            The U-value corresponding to the set of measurements in x
+        u2 : float
+            The U-value corresponding to the set of measurements in y
+
+    Notes
+    -----
+    Exact tests should be used for smaller sample sizes. Concretely, as
+    len(x) and len(y) increase to and beyond 8, the distribution of U differs
+    negligibly from the normal distribution[1]_. The default behavior of this
+    test is to calculate the number of possible sequences for inputs of
+    length(x) and length(y), and to do an exact calculation if the number of
+    possible combinations is <100000. The default behavior may be overridden
+    with the use_exact flag.
+
+    If an exact test is not performed, the U-distribution is approximated as a
+    normal distribution.
+
+    This test corrects for ties and by default uses a continuity correction
+    when approximating the U statistic distribution.
+
+    The reported p-value is for a two-sided hypothesis. To get the one-sided
+    p-value set alternative to 'greater' or 'less' (default is 'two-sided').
+
+    For Mann-Whitney-Wilcoxon tests, the reported U statistic is the U used to
+    test the hypothesis. The u1 and u2 statistics are returned as well,
+    corresponding to x and y, respectively.
+
+
+    .. versionadded:: 0.17.0
+
+    References
+    ----------
+    .. [1] H.B. Mann and D.R. Whitney, "On a test of whether one of two random
+           variables is stochastically larger than the other", The Annals of
+           Mathematical Statistics, Vol. 18, pp. 50-60, 1947.
+    .. [2] http://en.wikipedia.org/wiki/Mann-Whitney_U_test
+    """
+    x = asarray(x)
+    y = asarray(y)
+    if len(np.shape(x)) > 1 or len(np.shape(y)) > 1:
+        raise ValueError('Expected 1-d arrays.')
+    n1 = len(x)
+    n2 = len(y)
+    if exact == 'auto':
+        exact = ((n1 < 10 or n2 < 10) and n1 + n2 < 100000
+                 and -np.log(n1 + n2 + 1) - special.betaln(n1 + 1, n2 + 1) < np.log(100000))
+    ranked = rankdata(np.concatenate((x,y)))
+    rankx = ranked[0:n1]       # get the x-ranks
+    T = tiecorrect(ranked)
+    if T == 0:
+        raise ValueError('All numbers are identical')
+    if alternative not in ('two-sided', 'less', 'greater'):
```

Comment:
Style nitpick: input validation best be done before any other calculations (rationale: why ranking x and y if it's failing anyway)

Reply:
Implemented in `mannwhitneyu2`.